src/utils/resolveMessage.ts

```diff
@@ -0,0 +1,97 @@
+import {
+  ActionRowBuilder,
+  ButtonBuilder,
+  ButtonStyle,
+  Colors,
+  EmbedBuilder,
+  type Message,
+  ThreadChannel,
+} from 'discord.js';
+
+function findFirstImage(message: Message): string | undefined {
+  const result = message.attachments.find((attach) => {
+    return attach.contentType?.startsWith('image/');
+  });
+  if (result == undefined) {
+    return undefined;
+  } else {
+    return result.url;
+  }
+}
+
+export async function expandDiscordLink(message: Message): Promise<void> {
+  const re =
+    /(https?:\/\/)?(?:canary\.|ptb\.)?discord(?:app)?\.com\/channels\/(?<server_id>\d+)\/(?<channel_id>\d+)\/(?<message_id>\d+)/g;
+  let execResult = re.exec(message.content);
+
+  while (!(execResult == null || execResult.groups == undefined)) {
+    if (execResult.groups.server_id != message.guildId) {
+      continue; // do not let the bot leak messages from one server to another
```

Comment:
doesn't this cause an infinite loop, as execResult will still be non-null and still have groups?

Reply:
This doesn't cause an infinite loop because the /g specifier means that repeatedly executing the regex 'moves' the starting location for searching for regex.

You can see this happening in this code example:
```js
let pattern = /a\d/g;
let str = "a1a2a3";
console.log(pattern.exec(str)); // returns a1
console.log(pattern.exec(str)); // returns a2
console.log(pattern.exec(str)); // returns a3
console.log(pattern.exec(str)); // returns null
```

Reply:
Yes, but you don't exec it again here. You enter the while loop, check if the guild id differs and then just restart the loop and check again, restart, check again,...